src/golibmc.go

```diff
@@ -184,61 +329,253 @@ func New(servers []string, noreply bool, prefix string, hashFunc int, failover b
+		C.client_config(cn._imp, ConnectTimeout, client.connectTimeout)
+	}
+	return &cn, nil
 }
 
-// SimpleNew to create a memcached client with default params
-func SimpleNew(servers []string) (client *Client) {
-	noreply := false
-	prefix := ""
-	hashFunc := HashCRC32
-	failover := false
-	disableLock := false
-	return New(
-		servers, noreply, prefix, hashFunc,
-		failover, disableLock,
-	)
+func (client *Client) putConn(cn *conn, err error) error {
+	client.lk.Lock()
+	if isBadConnErr(err) || !client.putConnLocked(cn, nil) {
+		client.lk.Unlock()
+		err1 := cn.quit()
+		if err1 != nil {
+			log.Printf("Failed cn.quit: %v", err1)
+		}
+		return err
+	}
+	client.lk.Unlock()
+	return err
 }
 
-func finalizer(client *Client) {
-	C.client_destroy(client._imp)
+func (client *Client) putConnLocked(cn *conn, err error) bool {
+	if client.closed {
+		return false
+	}
+	if client.maxOpen > 0 && client.maxOpen < client.numOpen {
+		return false
+	}
+	if len(client.connRequests) > 0 {
+		var req chan connRequest
+		var reqKey uint64
+		for reqKey, req = range client.connRequests {
+			break
+		}
+		delete(client.connRequests, reqKey)
+		req <- connRequest{
+			conn: cn,
+			err:  err,
+		}
+	} else {
+		client.freeConns = append(client.freeConns, cn)
+		client.startCleanerLocked()
+	}
+	return true
 }
 
-func (client *Client) lock() {
-	if !client.disableLock {
+func (client *Client) conn(ctx context.Context) (*conn, error) {
+	cn, err := client._conn(ctx, true)
+	if err == nil {
+		return cn, nil
+	}
+	if err == ErrBadConn {
+		return client._conn(ctx, false)
+	}
+	return cn, err
+}
+
+func (client *Client) _conn(ctx context.Context, useFreeConn bool) (*conn, error) {
+	client.lk.Lock()
+	if client.closed {
+		client.lk.Unlock()
+		return nil, ErrMemcachedClosed
+	}
+	// Check if the context is expired.
+	select {
+	default:
+	case <-ctx.Done():
+		client.lk.Unlock()
+		return nil, ctx.Err()
+	}
+	lifetime := client.maxLifetime
+
+	var cn *conn
+	numFree := len(client.freeConns)
+	if useFreeConn && numFree > 0 {
+		cn = client.freeConns[0]
+		copy(client.freeConns, client.freeConns[1:])
+		client.freeConns = client.freeConns[:numFree-1]
+		client.lk.Unlock()
+		if cn.expired(lifetime) {
+			cn.quit()
+			return nil, ErrBadConn
+		}
+		return cn, nil
+	}
+
+	if client.maxOpen > 0 && client.maxOpen <= client.numOpen {
+		req := make(chan connRequest, 1)
+		reqKey := client.nextRequest
+		client.nextRequest++
+		client.connRequests[reqKey] = req
+		client.lk.Unlock()
+
+		select {
+		// timeout
+		case <-ctx.Done():
+			// Remove the connection request and ensure no value has been sent
+			// on it after removing.
+			client.lk.Lock()
+			delete(client.connRequests, reqKey)
+			client.lk.Unlock()
+			select {
+			case ret, ok := <-req:
+				if ok {
+					client.putConn(ret.conn, ret.err)
+				}
+			default:
+			}
+			return nil, ctx.Err()
+		case ret, ok := <-req:
+			if !ok {
+				return nil, ErrMemcachedClosed
+			}
+			return ret.conn, nil
```

Comment:
should comment if here assume that ret.err is nil?

Reply:
Well... I made it return ret.err.
https://github.com/douban/libmc/pull/57/commits/c1060ece6e88dfe094348438714e2a1084809d42